EM-647

mjfortier opened the pull request:
Collections - Allow document manipulation before saving

- This PR finalizes both EM-647, and the half-finished EM-218 currently in dev. Alerts, validation, and saving/publishing have all been properly implmented. QA ticket description to follow assuming code passes review.

NickPhura commented on modules/collections/client/views/collection-edit.html:
```diff
@@ -112,36 +111,43 @@ <h1 class="view-title__info--title">
+          <div class="alert alert-danger" ng-if="validationFlags.documentsTabInvalid">
             <span class="alert__icon glyphicon glyphicon-exclamation-sign"></span>
-            <span class="alert__msg">A minimum of one (1) PUBLISHED Main Document is required before this Collection can be published.</span>
+            <span class="alert__msg">A minimum of one (1) PUBLISHED Main Document is required for a Collection can be published.</span>
```
Some grammer weirdness

mjfortier replied:
Thank you, I meant to improve it but made it worse. Will fix immediately.

NickPhura commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -159,145 +164,210 @@ collectionModules.controller('CollectionEditCtrl',
+              }
+            })
+            .catch(function() {
+              AlertService.error('"' + $scope.collection.displayName + '" was unable to save successfully.', 4000, true);
```
I know this was here before, but "was unable to save successfully" reads a bit funny to me.
Do our UX folk have any thoughts?

RandyInVictoria replied:
Pretty sure that IS from our UX people lol

NickPhura commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -159,145 +164,210 @@ collectionModules.controller('CollectionEditCtrl',
+              AlertService.success('"' + $scope.collection.displayName + '" and documents were saved successfully.', 4000, true);
+              $scope.goToEdit();
+            }, function() {
+              AlertService.error('"' + $scope.collection.displayName + '" documents could not be saved.', 4000, true);
```
Related to my above comment: this seems more in line with what I would expect.

NickPhura commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -159,145 +164,210 @@ collectionModules.controller('CollectionEditCtrl',
+            })
+            .catch(function() {
+              // User cancels delete, or error is thrown.
+              AlertService.error('"'+ $scope.collection.displayName +'" was not deleted.', 4000, true);
```
Same comment again: this is a third style of error message.  Seems like we should unify what our messaging is.
Maybe this is for another story though.

mjfortier replied:
I've decided to go with this particular wording, and changed the rest to match it. Seems concise enough.

mjfortier commented:
Thanks @NickPhura , I've addressed those concerns and re-pushed to my repo. I also added the check to prevent duplicate collection names from existing, as I must have overlooked that while implementing the validation checks.

esune commented on modules/documents/client/directives/documents.manager.link.directive.js:
```diff
@@ -248,19 +248,14 @@ angular.module('documents')
               self.ok = function () {
                 // return the data in the selected list...
-                $uibModalInstance.close({linked: self.linkedFiles, source : self.sourceDocs[self.docType], docType : self.docType});
+                $uibModalInstance.close({linked: self.linkedFiles, docType : self.docType/*, source : self.sourceDocs[self.docType]*/});
```
Any reason to keep a commented-out parameter? It would be better if we removed it completely (same thing as line 258).

esune commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -1,139 +1,154 @@
+            updatedDocuments = _.without(updatedDocuments, doc);
+            // Remove them from the updated documents list, and alert the user.
+          })
+          AlertService.error('The following document(s) already exist in "' + (docType == 'main' ? 'Related' : 'Main') + ' Documents" and have not been added:' + errMsg, 7000, false);
```
Nit.: there is no need to pass the **false** parameter to the AlertService, as it is optional and will default to that value when omitted.

mjfortier replied:
Removed

esune commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -1,139 +1,154 @@
+      // Helper function to check if user has selected a document for both 'main' and 'other'.
+      $scope.findDuplicates = function(docs1, docs2) {
+        var duplicateFiles = [];
+        _.forEach(docs1, function(doc1) {
```
Nit.: what about using [_.union()](https://lodash.com/docs#union) for this purpose?

mjfortier replied:
_.union() wouldn't work, because I want an intersection. I've just learned, however, that lodash has a _.intersection() function. I'll try implementing it, but it uses Same Value Zero for comparison, which may or may not work with this.

esune commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -1,139 +1,154 @@
+            $scope.validationFlags.name = true;
+            AlertService.error('Project already contains a collection named "'+ $scope.collection.displayName +'".', 4000, true)
+          }
+          $scope.validationFlags.type = $scope.collection.type.length < 1 ? true : false;
```
No need to use a ternary operator here, `$scope.collection.type.length < 1` will already return true or false. The same should apply for the next flags as well.

mjfortier replied:
Good call. I was unsure how booleans are evaluated in JS, so I was playing it safe. It's one of those things I forgot to check when cleaning up my code at the end.

esune commented on modules/collections/client/controllers/collections.controllers.js:
```diff
@@ -159,145 +174,210 @@ collectionModules.controller('CollectionEditCtrl',
+            return $scope.submitDocs(false);
+          })
+          .then(function() {
+            $scope.collection.isPublished;
```
You may want to wait to set the "published" flag until the collection was actually published on the database (e.g.: when the promise returned by `CollectionModel.publishCollection($scope.collection._id)` is fulfilled).
Setting the state here could lead to weird situations IF the collection is not published for some reason.

mjfortier replied:
I'm actually going to remove this altogether, as I believe `CollectionModel.publishCollection($scope.collection._id);` sets it anyway
